fix leaky fake tensor on attribute assignment, support buffer assignment

avikchaudhuri opened the pull request:
In non-strict, assignment of attributes in a model causes their state to contain fake tensors post-tracing, which leads to incorrect results on running the exported model. We now error when this happens, asking the user to use buffers instead.
Next, we add support for assignment of buffers. The final values of the buffers turn into outputs of the graph. Since the buffers are already lifted as inputs and populated with the initial values when the model is run, this leads to a simple programming model where the driver of the model can feed the outputs back as inputs for successive runs.

Differential Revision: D55146852

bdhirsh commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -88,12 +88,49 @@ def aot_dispatch_base_graph(
+
+        def _map_assigned_buffer_to_proxy(_mod, name, buffer):
+            proxy_mode = None
+            for mode in torch.utils._python_dispatch._get_current_dispatch_mode_stack():
```
nit: you can do this:
```
torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.PROXY)
```

avikchaudhuri replied:
Thanks. 

As for asserting that the proxy mode always exists: unfortunately we don't have that guarantee right now. Saw test failures involving closures, cond, submodules, maybe more...

avikchaudhuri replied:
More details here: https://github.com/pytorch/pytorch/pull/122337#discussion_r1560182916

bdhirsh commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -88,12 +88,49 @@ def aot_dispatch_base_graph(
+                assigned_buffers[name] = fake_to_proxy.name
+            return buffer
+
+        handle = torch.nn.modules.module.register_module_buffer_registration_hook(
```
Do you mind adding a comment here describing the situation? There's a lot going on in this function and can be tough to keep track of everything (without the comment, the use of `get_proxy_slot` here looks a little suspicious)

avikchaudhuri replied:
Sure.

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -956,6 +1013,9 @@ def forward(self, *args, **kwargs):
                     wrapped_mod, new_preserved_call_signatures, module_call_specs
                 ):
                     gm, sig = aot_export(wrapped_mod, args, kwargs=kwargs, **flags)
+                    for buf, name in gm._assigned_buffers.items():
```
nit:

Today, the contract for `aot_export_module` is that it returns a tuple of `[GraphModule, GraphSignature]` (where the GraphSignature metadata contains any info about the GraphModule that someone who wants to call it needs to know, like any calling convention changes or input mutations).

I think it would be nice to maintain that invariant, instead of maintaining some side-car extra metadata directly through the `GraphModule`. That would require adding more fields to the `GraphSignature` class here: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/schemas.py#L550 - but I could also buy that this is kind of heavyweight.

avikchaudhuri replied:
Yeah this was ugly, fixed it now by modifying the metadata at the point when buffer assignments are detected, and letting the signature computation do its thing based on the metadata. Didn't need any new fields.

avikchaudhuri commented:
@bdhirsh PTAL

bdhirsh commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -48,6 +49,13 @@ def _create_graph(f, args, *, aot_config: AOTConfig) -> torch.fx.GraphModule:
 
 
+def _root_module_for_non_strict_export(flat_fn):
+    if hasattr(flat_fn, "_orig_mod") and hasattr(flat_fn._orig_mod, "_export_root"):
```
I am a little worried we run the risk of dropping this when someone adds another layer of wrapping/tracing to AOTAutograd 🤔  (cc @tugsbayasgalan )

I guess the alternative to this though is the previous "stash info about buffer swapping on the FX graph, so we can plumb it back up to the logic at the aot_export_module top-level that generates the GraphSignature". I don't have much of a preference between that or this

avikchaudhuri replied:
So ideally we would keep a reference to the original module somewhere, and no matter how much wrapping we do it should still be there as a stable API. I just don't know where it is or whether it even exists.

avikchaudhuri replied:
Added comment to make this dependency explicit.

tugsbayasgalan commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -88,12 +96,68 @@ def aot_dispatch_base_graph(
+            proxy_mode = torch._C._get_dispatch_mode(
+                torch._C._TorchDispatchModeKey.PROXY
+            )
+            if proxy_mode is not None and isinstance(
```
Hmm why would proxy mode be none here?

avikchaudhuri replied:
@bdhirsh asked the same question. I'm not sure why it can be none, but if I assert it is always non-none then I get various test failures. Let me make a catalog of those as a comment before landing.

avikchaudhuri replied:
Here's the list of errors on `fbcode//caffe2/test:test_export` when I assert: https://www.internalfb.com/intern/testinfra/testconsole/testrun/4785074831210563/

avikchaudhuri replied:
Here's a partial stack trace for one of the errors related to `cond` (which triggers `torch.compile(fullgraph=True)` which then seems to call `register_buffer` for some kind of setup?):

```
 File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/caffe2/test/export/test_export.py", line 3676, in forward
    return torch.cond(x.shape[0] > 4, true_fn, false_fn, [x])
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_higher_order_ops/cond.py", line 138, in cond
    return torch.compile(cond_op, backend="eager", fullgraph=True)(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/eval_frame.py", line 410, in _fn
    return fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 966, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state, skip=1)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 404, in _convert_frame_assert
    return _compile(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_utils_internal.py", line 279, in wrapper_function
    return StrobelightCompileTimeProfiler.profile_compile_time(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/caffe2/fb/strobelight/compile_time_profiler.py", line 89, in profile_compile_time
    return func(*args, **kwargs)
  File "/usr/local/fbcode/platform010/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 691, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/utils.py", line 265, in time_wrapper
    r = func(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 546, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/bytecode_transformation.py", line 1103, in transform_code_object
    transformations(instructions, code_options)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 168, in _fn
    return fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 508, in transform
    tracer.run()
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 2198, in run
    super().run()
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 848, in run
    while self.step():
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 758, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 2334, in RETURN_VALUE
    self._return(inst)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 2319, in _return
    self.output.compile_subgraph(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/output_graph.py", line 961, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/usr/local/fbcode/platform010/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/output_graph.py", line 1142, in compile_and_call_fx_graph
    gm = _make_graph_module(root, self.graph)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/fx/_lazy_graph_module.py", line 58, in _make_graph_module
    return graph_module_cls(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/fx/graph_module.py", line 400, in __init__
    _copy_attr(root, self, node.target)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/fx/graph_module.py", line 234, in _copy_attr
    to_module.register_buffer(field, orig)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/nn/modules/module.py", line 519, in register_buffer
    output = hook(self, name, tensor)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py", line 110, in _map_assigned_buffer_to_proxy
    assert proxy_mode is not None, f"{type(_mod)}.{name} = {type(buffer)}"
```

avikchaudhuri replied:
Here's another stack trace for a non-cond test: https://www.internalfb.com/intern/testinfra/diagnostics/4785074831210563.281475113795226.1712792368/

This one for `strict_mode_op`. So a HOP thing?

avikchaudhuri replied:
OK I'm going to try to narrow down when the hook is supposed to fire (basically only on the root module of export) to make stronger assertions. 🙏

avikchaudhuri committed:
fix leaky fake tensor on attribute assignment, support buffer assignment (#122337)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/122337

In non-strict, assignment of attributes in a model causes their state to contain fake tensors post-tracing, which leads to incorrect results on running the exported model. We now error when this happens, asking the user to use buffers instead.

Next, we add support for assignment of buffers. The final values of the buffers turn into outputs of the graph. Since the buffers are already lifted as inputs and populated with the initial values when the model is run, this leads to a simple programming model where the driver of the model can feed the outputs back as inputs for successive runs.

Test Plan: added tests

Reviewed By: tugsbayasgalan, bdhirsh

Differential Revision: D55146852
